File: Squirrel2/Squirrel2/Models/CollectionEntrySuggestion.swift
import Foundation
import FirebaseFirestore

struct CollectionEntrySuggestion: Identifiable {
    enum SuggestionType: String {
        case fetchProduct = "fetch_product_suggestion"
        case recipe = "recipe_suggestion"
        case unknown
    }

    enum Status: String {
        case pending
        case completed
        case failed
    }

    struct ProductSuggestion: Identifiable {
        let id = UUID()
        let merchant: String
        let title: String
        let price: String?
        let url: URL?
        let notes: String?
    }

    let id: String
    let collectionEntryId: String?
    let entryId: String?
    let collectionId: String?
    let userId: String
    let type: SuggestionType
    let status: Status
    let payload: [String: Any]
    let error: String?
    let metadata: [String: Any]?
    let createdAt: Date
    let updatedAt: Date

    init?(document: DocumentSnapshot) {
        guard let data = document.data(),
              let userId = data["userId"] as? String,
              let typeRaw = data["type"] as? String,
              let statusRaw = data["status"] as? String else {
            return nil
        }

        self.id = document.documentID
        self.collectionEntryId = data["collectionEntryId"] as? String
        self.entryId = data["entryId"] as? String
        self.collectionId = data["collectionId"] as? String
        self.userId = userId
        self.type = SuggestionType(rawValue: typeRaw) ?? .unknown
        self.status = Status(rawValue: statusRaw) ?? .pending
        self.payload = data["payload"] as? [String: Any] ?? [:]
        self.error = data["error"] as? String
        self.metadata = data["metadata"] as? [String: Any]

        if let createdTs = data["createdAt"] as? Timestamp {
            self.createdAt = createdTs.dateValue()
        } else {
            self.createdAt = Date()
        }

        if let updatedTs = data["updatedAt"] as? Timestamp {
            self.updatedAt = updatedTs.dateValue()
        } else {
            self.updatedAt = self.createdAt
        }
    }

    var productSuggestions: [ProductSuggestion] {
        guard type == .fetchProduct,
              let rawSuggestions = payload["suggestions"] as? [[String: Any]] else {
            return []
        }

        return rawSuggestions.compactMap { item in
            guard let title = item["title"] as? String else { return nil }
            let merchant = (item["merchant"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
            let merchantValue = merchant?.isEmpty == false ? merchant! : "Unknown retailer"
            let price = (item["price"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
            let notes = (item["notes"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)

            var url: URL? = nil
            if let urlString = item["url"] as? String,
               let candidate = URL(string: urlString),
               candidate.scheme?.hasPrefix("http") == true {
                url = candidate
            }

            return ProductSuggestion(
                merchant: merchantValue,
                title: title,
                price: price?.isEmpty == false ? price : nil,
                url: url,
                notes: notes?.isEmpty == false ? notes : nil
            )
        }
    }
}

File: backend/src/models/CollectionEntrySuggestion.js
const { firestore, admin } = require('../config/firebase');

class CollectionEntrySuggestion {
  constructor(data = {}) {
    this.id = data.id || null;
    this.collectionEntryId = data.collectionEntryId || null;
    this.entryId = data.entryId || null;
    this.collectionId = data.collectionId || null;
    this.userId = data.userId || null;
    this.type = data.type || null;
    this.status = data.status || 'pending';
    this.payload = data.payload || null;
    this.error = data.error || null;
    this.metadata = data.metadata || {};
    this.createdAt = data.createdAt || new Date();
    this.updatedAt = data.updatedAt || new Date();
  }

  static collection() {
    return firestore.collection('collection_entry_suggestions');
  }

  static async create(data) {
    const suggestion = new CollectionEntrySuggestion(data);
    const docRef = await this.collection().add({
      collectionEntryId: suggestion.collectionEntryId,
      entryId: suggestion.entryId,
      collectionId: suggestion.collectionId,
      userId: suggestion.userId,
      type: suggestion.type,
      status: suggestion.status,
      payload: suggestion.payload,
      error: suggestion.error,
      metadata: suggestion.metadata,
      createdAt: admin.firestore.Timestamp.fromDate(suggestion.createdAt),
      updatedAt: admin.firestore.Timestamp.fromDate(suggestion.updatedAt)
    });

    suggestion.id = docRef.id;
    return suggestion;
  }

  static async findById(id) {
    const doc = await this.collection().doc(id).get();
    if (!doc.exists) return null;

    const data = doc.data();
    return new CollectionEntrySuggestion({
      id: doc.id,
      ...data,
      createdAt: data.createdAt?.toDate?.() || data.createdAt || new Date(),
      updatedAt: data.updatedAt?.toDate?.() || data.updatedAt || new Date()
    });
  }

  async update(fields = {}) {
    if (!this.id) {
      throw new Error('Cannot update CollectionEntrySuggestion without ID');
    }

    const updates = { ...fields, updatedAt: admin.firestore.Timestamp.fromDate(new Date()) };
    await CollectionEntrySuggestion.collection().doc(this.id).update(updates);

    Object.assign(this, fields, { updatedAt: new Date() });
    return this;
  }
}

module.exports = CollectionEntrySuggestion;
